python/pyspark/sql/types.py

```diff
@@ -196,7 +199,9 @@ def toInternal(self, dt):
     def fromInternal(self, ts):
         if ts is not None:
             # using int to avoid precision loss in float
-            return datetime.datetime.fromtimestamp(ts // 1000000).replace(microsecond=ts % 1000000)
+            y, m, d, hh, mm, ss, _, _, _ = (time.gmtime(ts // 1000000) if _is_utc
+                                            else time.localtime(ts // 1000000))
+            return datetime.datetime(y, m, d, hh, mm, ss, ts % 1000000)
```

Comment:
I'd explain how the current change is equivalent to `fromtimestamp` in the PR description.

https://github.com/python/cpython/blob/018d353c1c8c87767d2335cd884017c2ce12e045/Lib/datetime.py#L1425-L1458

Reply:
I think the only difference is this `ss = min(ss, 59)`

Reply:
Could we have more details in the PR description? I can follow what it fixes but I believe we should better describe details for other guys to follow the history in the future. To me, it has been always hard to track such changes and ensure the root cause of a bug.

Reply:
BTW, the change itself seems fine to me.

Reply:
I added some description and support for leap seconds